api/openapi-spec/swagger.json

```diff
@@ -18365,7 +18365,7 @@
     }
    },
    "v1.VirtualMachineInstancePreset": {
-    "description": "VirtualMachineInstancePreset defines a VMI spec.domain to be applied to all VMIs that match the provided label selector More info: https://kubevirt.io/user-guide/virtual_machines/presets/#overrides",
+    "description": "Deprecated for removal in v2, please use VirtualMachineInstanceType and VirtualMachinePreference instead.\n\nVirtualMachineInstancePreset defines a VMI spec.domain to be applied to all VMIs that match the provided label selector More info: https://kubevirt.io/user-guide/virtual_machines/presets/#overrides",
```

Comment:
i'm all for removing presets. I'm not sure what `v2` refers to here. Is the expectation that we'll increment our core api to v2 for this deprecation? I thought we'd just drop presets eventually and let that be it while still in v1.

Reply:
Yeah I don't think we have a documented policy in KubeVirt but reading the k8s docs [1] suggested that `v2` would be the earliest we could drop it. 

[1] https://kubernetes.io/docs/reference/using-api/deprecation-policy/#rest-resources-aka-api-objects

Reply:
@davidvossel thoughts on this? Apologies if I've misunderstood something here.

Reply:
We don't really have plans for a `v2` api right now. We also dont' really have a policy for how to drop `v1` apis.

My suggestion would be to propose a policy for API deprecation here [1] and for us to put something in writing that works for us before removing presets. Looking at this a little closer, we're kind of in an unfortunate place since we graduated presets to `v1`.

It would be helpful to find any examples of how the k8s core handled deprecating something that was `v1` as well. I'm not sure if it has ever been done. For example, PodPresets was still a beta i believe.

1. https://github.com/kubevirt/kubevirt/blob/main/docs/deprecation.md
2. https://github.com/kubernetes/kubernetes/pull/94090

Reply:
Thanks @davidvossel, comments below;

> We don't really have plans for a `v2` api right now. We also dont' really have a policy for how to drop `v1` apis.

Without a policy I would assume we would default to the k8s policy I listed in my last comment that highlights the need to wait until a newer `v2` version. The fact we don't have plans for such a version at the moment shouldn't stop us from starting the deprecation process IMHO.

> My suggestion would be to propose a policy for API deprecation here [1] and for us to put something in writing that works for us before removing presets. Looking at this a little closer, we're kind of in an unfortunate place since we graduated presets to `v1`.

Can we differ from k8s here without impacting our project status with the CNCF and more importantly confusing end users by say removing parts of a `v1` API after N KubeVirt releases instead of N releases with a `v2` API? Happy to look at this if not but I'm slightly worried about diverging too much from k8s in terms of API deprecation policies. 

Also yeah the fact presets and most of our core API jumped from `v1alpha3` to `v1` is a little unfortunate. We likely need to deprecate and later remove the `v1alpha3` version sometime soon.

> It would be helpful to find any examples of how the k8s core handled deprecating something that was `v1` as well. I'm not sure if it has ever been done. For example, PodPresets was still a beta i believe.

Yeah indeed, there's a handy readout available below:

https://kubernetes.io/docs/reference/using-api/deprecation-guide/#removed-apis-by-release

The closest thing I can find is `autoscaling/v1` and `autoscaling/v2` but the former isn't deprecated yet while the latter has recently deprecated older `autoscaling/v2beta1` and `autoscaling/v2beta2` versions.

Reply:
>Can we differ from k8s here without impacting our project status with the CNCF and more importantly confusing end users by say removing parts of a v1 API after N KubeVirt releases instead of N releases with a v2 API? Happy to look at this if not but I'm slightly worried about diverging too much from k8s in terms of API deprecation policies.

Umm, I'm not finding any precedence for removing a v1 api in the k8 ecosystem. I think you're right that we'll need to eventually move to a v2 in order to drop presets.

lets make this a topic in the community call and try to arrive at a consensus here for what our policy will be.

Reply:
> Umm, I'm not finding any precedence for removing a v1 api in the k8 ecosystem. I think you're right that we'll need to eventually move to a v2 in order to drop presets.

Yes seems like I mixed it up with something when I wrote my initial comment about the removal after 1 year.

> lets make this a topic in the community call and try to arrive at a consensus here for what our policy will be.

I can't take part of community calls at the moment. I don't have any strong opinion though. My rought thoughtline which I would favour right now:

 * In general fall in line with the k8s policy
 * Maybe have an exception for pod presets (they are pretty unique) and let us remove them this time
 * For future features it would be easier to fall in line with the k8s policy, because they would have their own api groups and would not impact all core resources. They can easily go to v2, v3 and so forth.

Reply:
Thanks both, I can raise this on the call today but I personally don't think it's going to alter the requirement to wait until `v2` before this can be removed.

Reply:
So I also have a conflict this week with the meeting, I'll raise this in the ML thread.

Reply:
Nothing of note on the ML, @davidvossel shall we just go ahead with the `v2` wording and start the deprecation?

[1] https://groups.google.com/g/kubevirt-dev/c/eM7JaDV_EU8

Reply:
yes, practically, i think this is how we need to communicate this. We need to begin discouraging the usage of presets in favor of instancetypes/preferences. It's a little awkward that V2 has no timeline right now, but i think you've worded this as accurately as we can right now